proxy/src/transparency/tcp.rs

```diff
@@ -71,14 +73,177 @@ impl Proxy {
         let fut = connect.connect()
             .map_err(|e| debug!("tcp connect error: {:?}", e))
             .and_then(move |tcp_out| {
-                let (in_r, in_w) = tcp_in.split();
-                let (out_r, out_w) = tcp_out.split();
-
-                copy(in_r, out_w)
-                    .join(copy(out_r, in_w))
-                    .map(|_| ())
+                Duplex::new(tcp_in, tcp_out)
                     .map_err(|e| debug!("tcp error: {}", e))
             });
         Box::new(fut)
     }
 }
+
+/// A future piping data bi-directionally to In and Out.
+struct Duplex<In, Out> {
+    in_buf: CopyBuf,
+    in_eof: bool,
+    in_io: In,
+    in_shutdown: bool,
+    out_buf: CopyBuf,
+    out_eof: bool,
+    out_io: Out,
+    out_shutdown: bool,
+}
+
+/// A buffer used to copy bytes from one IO to another.
+///
+/// Keeps read and write positions.
+struct CopyBuf {
+    // In linkerd-tcp, a shared buffer is used to start, and an allocation is
+    // only made if NotReady is found trying to flush the buffer. We could
+    // consider making the same optimization here.
+    buf: Box<[u8]>,
+    read_pos: usize,
+    write_pos: usize,
+}
+
+impl<In, Out> Duplex<In, Out>
+where
+    In: AsyncRead + AsyncWrite,
+    Out: AsyncRead + AsyncWrite,
+{
+    fn new(in_io: In, out_io: Out) -> Self {
+        Duplex {
+            in_buf: CopyBuf::new(),
+            in_eof: false,
+            in_io,
+            in_shutdown: false,
+            out_buf: CopyBuf::new(),
+            out_eof: false,
+            out_io,
+            out_shutdown: false,
+        }
+    }
+
+    /// Returns true if the Duplex is completely done.
+    ///
+    /// The future may finish earlier if one side has an error.
+    fn is_done(&self) -> bool {
+        self.in_eof
+            && !self.in_buf.has_remaining()
+            && self.out_eof
+            && !self.out_buf.has_remaining()
+    }
+}
+
+impl<In, Out> Future for Duplex<In, Out>
+where
+    In: AsyncRead + AsyncWrite,
+    Out: AsyncRead + AsyncWrite,
+{
+    type Item = ();
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // this is macro because we don't want to actually split the IO, as
+        // that requires a futures 'TryLock'. Instead, so that we can skip
+        // that lock, the Duplex holds the original IO, and this macro
+        // prevents duplication of code.
```

Comment:
I admit I don't quite understand all of this yet. However, can it be refactored to be more understandable?

In particular, I get that there might be a reason why some code here should be in a macro. But, is there any way to factor out a large amount of this code to be a regular function(s) that is (are) used by the macro, so that we can see clearly in the macro definition what is really motivating the use of a macro and avoiding the use of `TryLock`? Right now it is really unclear (at least to me, somebody who doesn't deal with these things often). In general I think it's good to pull out as much code from macros as possible.